File: gyakorlas2.py
#ebban taroljuk a usereket
users = []

class User():
    def __init__(self, email, name, premium, password):
        self.email = email
        self.name = name
        self.premium = premium
        self.password = password
        self.signincount = 0

    def __str__(self):
        return "The users email is {} the name is {} the number of sign in is {} and is a {} user.".format(self.email,self.name,self.signincount,self.premium)

    def setEmail(self, email):
        self.email = email

    def getEmail(self):
        return self.email

    def setName(self, name):
        self.name = name

    def getName(self):
        return self.name

#nem szetteljuk, mert belulrol allitja magat
    def setSignInCount(self, signincount):
        self.signincount = signincount

    def getSignInCount(self):
        return self.signincount

    def setPremium(self, premium):
        self.premium = premium

    def getPremium(self):
        return self.premium

    def setPassword(self, password):
        self.password = password

    
#nem akarjuk hogy barki lassa
    def getPassword(self):
        return self.password

    def delete(self):
        users.remove(user)
                
    def isPasswordValid(self, password):
        if self.getPassword() == password: 
            return True
        else:
            return False

    @classmethod
    def register(cls, email, name, premium, password):
        user = cls(email, name, premium, password)
        users.append(user)

    @classmethod
    def findByEmail(cls, email):
        for user in users:
            if user.getEmail() == email:
                return user
        return None

    @classmethod
    def findAll(cls):
        return users
    

    @classmethod
    def login(cls, email, password):
        user = cls.findByEmail(email)

        if user is None:
            print("email or password not valid")
            return None
        
        if user.isPasswordValid(password):
            incrementedSignInCount = user.getSignInCount() + 1
            user.setSignInCount(inceremtedSignInCount)

            print("Welcome {}".format(email))
            return user
        else:
            print("email or password not valid")
                   





        
# User instance
tamas = User("example@examle.com", "Lajos", True, '1234')

# User instance string representation
print(str(tamas))

# User instance getter
tamas.getName()


# User methods
User.register("emailmarhajo@example.com", "geza", True, "25")
User.register("test@test.com", "testname", False, "32")

allUsers = User.findAll()
print(allUsers)

# User successfully found (user instance)
User.findByEmail("emailmarhajo@example.com")

# User was not found found (None)
User.findByEmail("nincsilyenemail@example.com")

user = User.login('emailmarhajo@example.com', "25")
user.isPasswordValid("123")
user.delete()
print(allUsers)

File: gyakorlas2.rb
$users = []

class User
    attr_accessor :email, :name, :premium
    attr_writer :password
    attr_reader :signincount

    def initialize(email, name, premium, password)
        @email = email
        @name = name
        @premium = premium
        @password = password
        @signincount = 0
    end

    def to_s
        "The users email is #{@email} the name is #{@name} the number of sign in is #{@premium} and is a #{@password} user."
    end

    def delete
        $users - [self]
    end

    def password_valid?(password)
        @password == password
    end

    class << self
        def register(email, name, premium, password)
            user = new(email, name, premium, password)
            $users << user
        end
    end
end

# objektum letrehozasa
tamas = User.new('example@examle.com', "Lajos", true, '1234')

#objektum string formaban a to_s miatt
puts tamas

# email kilvasasa a meglevo objektumbol, es email beleirasa az objektumba
puts tamas.email
tamas.email = 'new@example.com'

# sign in count kivulrol olvashato, de nem irhato
puts tamas.signincount

# ez hibat fog dobni, mert nem irhato
# tamas.signincount = 1

# password kivulrol irhato, de nem olvashoato
tamas.password = 'new pssword'

# ez hibat fog dobni, mert nem olvashato
# tamas.password

#user registration
User.register('emailmarhajo@example.com', 'geza', true, '25')

# users tomb kiirasa, ahova beleregisztratuk az uj usert
puts $users


tamas.delete

puts $users

puts tamas.password_valid?('new pssword')